File: csharp/FunctionalRefactoring/Models/CartId.cs
using System;

namespace FunctionalRefactoring.Models
{
    public class CartId
    {
        public String Value { get; }

        public CartId(String value) => Value = value;

        Boolean Equals(CartId other) => String.Equals(Value, other.Value);

        public override Boolean Equals(Object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != typeof(CartId)) return false;
            return Equals((CartId)obj);
        }

        public override Int32 GetHashCode() => Value != null ? Value.GetHashCode() : 0;

        public static Boolean operator ==(CartId left, CartId right) => Equals(left, right);

        public static Boolean operator !=(CartId left, CartId right) => !Equals(left, right);
    }
}

File: java/src/main/java/org/functionalrefactoring/models/CartId.java
package org.functionalrefactoring.models;

import java.util.Objects;

public class CartId {
    public final String value;

    public CartId(String value) {
        this.value = value;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CartId cartId = (CartId) o;
        return Objects.equals(value, cartId.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
}
